### ISSUE TITLE
Segmentation fault when closing a dll which uses openmp

### ISSUE THREAD

**Issue by longqian95** [JuliaLang/julia#10938]
Here is testing code in linux

test.cpp:

``` C++
extern "C" int test(int n)
{
    int s=0;
    #pragma omp parallel for reduction (+:s)
    for(int i=0;i<n;++i) s+=i;
    return s;
}
```

build dll:

```
gcc -shared -fPIC -fopenmp test.cpp
```

Julia code causing segmentation fault:

``` julia
h=dlopen("a.out"); ccall(dlsym(h,:test),Int,(Int,),3); dlclose(h)
```

Note:
Both 0.3.7 and the master version(0.4.0-dev+4392) have this problem. If there are other statements between `ccall` and `dlclose`, such as `println(ccall(dlsym(h,:test),Int,(Int,),3)); dlclose(h)`,  sometimes there is no problem. If openmp is not used, there is no problem. If the dll is already loaded, such as by `ccall((:test,"a.out"),Int,(Int,),3)`, the above code does not cause fault.

Version info:
Julia Version 0.3.7
Commit cb9bcae\* (2015-03-23 21:36 UTC)
Platform Info:
  System: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU           X5675  @ 3.07GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libopenlibm
  LLVM: libLLVM-3.3

**Comment by vtjnash** [comment]
http://stackoverflow.com/questions/5200418/destroying-threads-in-openmp-c

tl;dr it's probably not safe to call dlclose on an openmp binary

**Comment by tkelman** [comment]
quoth @JeffHammond:
> I've been discussing these issues in the OpenMP interop working group for a year or so...

Any directions coming out of these discussions yet?

**Comment by longqian95** [comment]
How about change `dlclose(h)` to `@async begin sleep(0.01);  dlclose(h); end`. Is it safe?

I care about this because I use the following macro for debugging dll:
```julia
macro dynamic(exp)
    if exp.head!=:ccall || exp.args[1].head!=:tuple
        error("unsupported ccall expression")
    end
    func=esc(exp.args[1].args[1])
    dl=esc(exp.args[1].args[2])
    args=map(esc,exp.args[2:end])

    return quote
        let
            const h=dlopen(string($(dl)))
            try
                ccall(dlsym(h,string($func)),$(args...))
            finally
                @async begin
                    sleep(0.01)
                    dlclose(h)
                end
            end
        end
    end
end
```
For normal `ccall(...)`, I have to restart Julia every time after I modified dll. With this macro, I simply change `ccall(...)` to `@dynamic ccall(...)` then I can modify and test dll freely. Previous, I just put `dlcose(h)` in the try-finally clause and found it is not safe when using openmp.

**Comment by ihnorton** [comment]
I don't think that would be any safer; the issue is not with using sync/async on the Julia side, but rather that you need to ensure the threads are all safely cleaned up on the OpenMP side (which doesn't know anything about the Julia scheduler). There are a couple suggestions around for how to do that, manually, which you might try to implement in your own code and then ccall from Julia before `dlclose`ing.

http://bisqwit.iki.fi/story/howto/openmp/#CancellingOfThreadsAndBreakingOutOfLoops

http://stackoverflow.com/questions/3914264/openmp-is-there-a-way-for-a-thread-to-terminate-all-other-parallel-threads

**Comment by jeffhammond** [comment]
Can someone write a simple reproducer in C so I can bring it to the OpenMP community as a compelling reason to do what I have been asking for, which is to add an API call to shutdown the thread pool?  I will also use this example to motivate implementers to support it even if it's not in the standard.

**Comment by tkelman** [comment]
> Can someone write a simple reproducer in C so I can bring it to the OpenMP community

We should try to do this, seems kind of important. Looks like `dlopen` and `dlclose` in Julia are pretty direct wrappers around the C versions.

**Comment by AlexandreEichenberger** [comment]
Have a question for you guys; OpenMP can shut down threads, but what happens if another thread is also doing OpenMP? Would you want to shut down only inactive threads?  Or return a value to indicate that all threads could not be terminated because some where still in use? Or block until all of them are terminated?

**Comment by ihnorton** [comment]
@jeffhammond here is a minimal reproducer:
```
// gcc -o ompdldie ompdldie.c -ldl

#include <dlfcn.h>

#define OMPLIB "libgomp"

void main () {
    void* hdl = dlopen(OMPLIB, RTLD_LOCAL);
    dlclose(hdl);
}
```
```
isaiah@titan:/tmp/t2$ ./ompdldie 
Segmentation fault
```


(`mpich2                     3.0.4-6ubuntu1     amd64              Transitional dummy package`)

**Comment by jeffhammond** [comment]
Thanks!  I am amazed that this fails without even including an OpenMP construct.

**Comment by jeffhammond** [comment]
I don't know if this is sufficient to reproduce.  It doesn't segfault on Mac:
```
jrhammon-mac01:BUGS jrhammon$ icc -o ompdldie ompdldie.c -ldl
jrhammon-mac01:BUGS jrhammon$ ./ompdldie 
jrhammon-mac01:BUGS jrhammon$ echo $?
0
jrhammon-mac01:BUGS jrhammon$ gcc-4.9 -o ompdldie ompdldie.c -ldl
jrhammon-mac01:BUGS jrhammon$ ./ompdldie 
jrhammon-mac01:BUGS jrhammon$ echo $?
255
```

**Comment by ihnorton** [comment]
I have another version of the code using the OPs function, but I have no
idea if it will repro on OS X:

https://gist.github.com/ihnorton/500f6add042d8c7411f2

**Comment by ihnorton** [comment]
I'm using gcc 4.8.2 (on Ubuntu 14.04) in case that makes a difference.

**Comment by jeffhammond** [comment]
Thanks.  The more complicated test still runs fine for me.  I will test more on my Linux box when I get back to the office.

**Comment by ihnorton** [comment]
Running it repeatedly, I see that it actually succeeds about 30% of the time. Very strange.

**Comment by yuyichao** [comment]
AFAICT this is an upstream issue and it's not even an upstream we rely on so close.
